src/v/redpanda/request_auth.cc

```diff
@@ -0,0 +1,190 @@
+              "Unauthorized", ss::httpd::reply::status_type::unauthorized);
+        } else {
+            const auto& cred = cred_opt.value();
+            bool is_valid = security::scram_sha256::validate_password(
+              password, cred.stored_key(), cred.salt(), cred.iterations());
+            if (!is_valid) {
+                // User found, password doesn't match
+                vlog(
+                  logger.info,
+                  "Client auth failure: user '{}' wrong password",
+                  username);
+                throw ss::httpd::base_exception(
+                  "Unauthorized", ss::httpd::reply::status_type::unauthorized);
+            } else {
+                vlog(logger.trace, "Authenticated user {}", username);
+                const auto& superusers = config::shard_local_cfg().superusers();
+                auto found = std::find(
+                  superusers.begin(), superusers.end(), username);
+                bool superuser = (found != superusers.end()) || (!require_auth);
+                return request_auth_result(username, superuser);
+            }
+        }
+    } else if (!auth_hdr.empty()) {
+        throw ss::httpd::bad_request_exception(
+          "Unsupported Authorization method");
+    } else {
+        // No Authorization header: user is anonymous
+        if (require_auth) {
+            return request_auth_result(false, false);
+        } else {
+            return request_auth_result(true, true);
+        }
+    }
+}
+
+void request_auth_result::require_superuser() {
+    _checked = true;
+    if (!_superuser) {
+        vlog(
+          logger.info,
+          "Client authorization failure: {} is not a superuser",
+          _username);
+        throw ss::httpd::base_exception(
+          "Forbidden (superuser role required)",
+          ss::httpd::reply::status_type::forbidden);
+    }
+}
+
+void request_auth_result::require_authenticated() {
+    _checked = true;
+    if (!_authenticated) {
+        vlog(
+          logger.info,
+          "Client authorization failure: user must be authenticated");
+        throw ss::httpd::base_exception(
+          "Forbidden (authentication is required)",
+          ss::httpd::reply::status_type::unauthorized);
+    }
+}
+
+void request_auth_result::pass() { _checked = true; }
+
+/**
+ * It is important to protect against someone calling authenticate()
+ * but then not calling any of the authorization helpers: this indicates
+ * a request handler that might be unintentionally allowing unchecked
+ * access.
+ *
+ * This is a rare case of a throwing destructor.  It is made safe by
+ * checking if there is already an exception in flight first, and by
+ * knowing that all our member objects have nothrow destructors.
+ */
+request_auth_result::~request_auth_result() noexcept(false) {
+    if (!_checked && !std::current_exception()) {
+        vlog(
+          logger.error, "request_auth_result destroyed without being checked!");
```

Comment:
We could, but my intent here was that in the event of another exception in a request handler (between authenticate() and checking), it is okay that we didn't check.

If we logged this message while another exception is in flight, I think it could create a confusing situation where the log is complaining about something to do with auth, where the real exception is something totally different.